kotlin-redux/README.md

```diff
@@ -0,0 +1,48 @@
+## kotlin-redux
+
+Kotlin wrapper for Redux library. Major version number of this wrapper matches that of Redux itself.
+
+## Creating and using a store in Kotlin
+
+In Redux, the store is an object that contains a state, which could be anything. In Kotlin, this is 
+no different; just use RStore instead of createStore. Note that initializing the state is
+mandatory.
+
+```
+val myStore = RStore(::myReducer, someState)
+```
+Access the state with the store's `state` field.
+
+```
+myStore.state
+```
+In Kotlin, Actions can be written as classes. Just implement the `RAction` interface.
```

Comment:
Does that mean that they are non-serializeable?

Reply:
To make the actions type-safe, it uses a single internal 'real' redux action to dispatch the actions that you write in kotlin.

Reply:
I wonder if there is any way to keep both typesafety and serializeability

Reply:
Something like this is always an option:
```someAction = RAction(type = "sometype")```
but this would be ugly in a typesafe language like kotlin. If actions needed to be serialized, would it work to instance them like so?
```actionsToDispatch = listOf(SomeAction("someParameter"), ...)```
this would not capture the action in general, but would work for specific use cases of the action.

Reply:
Sorry, I'm not sure I'm following

My question is — if we're using checks like `is MyAction` in reducer, how will it work with deserialized data?

Reply:
Do you mean how would the parameters work?
```
data class AddToString(val string: String) : RAction

fun reducer(state: String, action: RAction) = when {
	is AddToString -> state + action.string // smart cast
	else -> state
}
```
is what I was thinking. I tested this, and it seems to work fine. I can change it though if it is a problem.

Reply:
If `action` is deserialized from json, `is AddToString` check will fail

Reply:
Related discussion:
https://github.com/reactjs/redux/issues/437

Reply:
Hmm... I tested it though, and it seems to work...
I think it's because we wrap it in an actual redux action, like so:
```
js {
	type = "WRAPPER_ACTION",
	action = SomeRAction()
}
```
It's a pure javascript object that has a more complex kotlin object as one of it's keys, that gets processed by the same actual redux reducer every time, which is the `wrapperReducer()` method that I wrote in the `RStore` class.

I realize that it would be preferable to use unique redux actions instead of a single one to execute all the fake redux actions that we write in kotlin, though...
I will try to think of a solution to this. Reflection, maybe?

Reply:
How exactly did you test that?

Reply:
I made an example state, and example reducer, and some example actions, and then used the store to dispatch them, and got the results by logging it. No errors occurred, and the actions did what they intended.

Reply:
So where's the serialization/deserialization part?

Reply:
The action class is instanced and stored under the `action` key in the wrapper action.

Reply:
Serialization means means converting to string, and deserialization converting from string back to object. It's needed for features like replaying actions

Reply:
So basically add a toJson method?

Reply:
Yes. And my concern is that after parsing the Json back, Kotlin type information will be lost

Reply:
It can be serialized into json via kotlinx.serialization. Also, to preserve type, a type key can be included in the json, which can then be checked for when converting the json back into a kotlin object.

Reply:
how should usages of `deserializeAction` look like?

Reply:
```
val jsonAction = "{\"type\":\"SomeAction\",\"someArg\":25}"
val action = deserializeAction<SomeAction>(jsonAction)
if (action != null) store.dispatch(action)
```

Reply:
How do you know which type to use when deserializing?

Reply:
My question is, how do I know that I should cast it to `SomeAction` not to `SomeOtherAction` if the string is only available at runtime

Reply:
```
val action = when (JSON.parse((StringSerializer to StringSerializer).map, jsonAction)["type"]) = {
	"SomeAction" -> deserializeAction<SomeAction>(jsonAction)
	"SomeOtherAction" -> deserializeAction<SomeOtherAction>(jsonAction)
	else -> null
}
```
works, where action is a nullable RAction.

Reply:
OK, please add it to readme